Userland/tee.cpp

```diff
@@ -0,0 +1,96 @@
+#include <AK/Vector.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+
+static Vector<int> collect_fds(int argc, char** argv, int start, bool aflag, bool* err)
+{
+    int oflag;
+    mode_t mode;
+    if (aflag) {
+        oflag = O_APPEND;
+        mode = 0;
+    } else {
+        oflag = O_CREAT | O_WRONLY | O_TRUNC;
+        mode = S_IROTH | S_IWOTH | S_IRGRP | S_IWGRP | S_IRUSR | S_IWUSR;
+    }
+
+    Vector<int> fds;
+    for (int i = start; i < argc; ++i) {
+        int fd = open(argv[i], oflag, mode);
+        if (fd < 0) {
+            perror("failed to open file for writing");
+            *err = true;
+        }
+        fds.append(fd);
+    }
+    fds.append(STDOUT_FILENO);
+    return fds;
+}
+
+static void copy_stdin(Vector<int>& fds, bool* err)
+{
+    for (;;) {
+        char buf[4096];
+        ssize_t nread = read(STDIN_FILENO, buf, sizeof(buf));
+        if (nread == 0)
+            break;
+        if (nread < 0) {
+            perror("read() error");
+            *err = true;
```

Comment:
Shouldn't we return here?

Reply:
Probably, but I wasn't able to find a specific section in the docs speaking to this behavior. If we fail to read from stdin though, things are probably done.

Reply:
How would we break out of this loop otherwise?

Reply:
Yes agreed